cloudinit/sources/helpers/hetzner.py

```diff
@@ -3,22 +3,44 @@
 #
 # This file is part of cloud-init. See LICENSE file for license information.
 
-from cloudinit import url_helper, util
+from typing import Optional
 
+from cloudinit import net, url_helper
 
-def read_metadata(url, timeout=2, sec_between=2, retries=30):
-    response = url_helper.readurl(
-        url, timeout=timeout, sec_between=sec_between, retries=retries
-    )
-    if not response.ok():
-        raise RuntimeError("unable to read metadata at %s" % url)
-    return util.load_yaml(response.contents.decode(), allowed=(dict, list))
 
+def skip_retry_on_empty_response(cause: url_helper.UrlError) -> bool:
+    """Returns False if cause.code is 204."""
+    return cause.code != 204
 
-def read_userdata(url, timeout=2, sec_between=2, retries=30):
-    response = url_helper.readurl(
-        url, timeout=timeout, sec_between=sec_between, retries=retries
-    )
-    if not response.ok():
-        raise RuntimeError("unable to read userdata at %s" % url)
-    return response.contents
+
+def get_metadata(
+    urls,
+    max_wait=120,
+    sec_between=2,
+    retries=30,
+    timeout=2,
+    sleep_time=2,
+    exception_cb=None,
+) -> tuple[str, bytes]:
```

Comment:
This is not compatible with the Python 3.8, which is still supported.

```suggestion
) -> Tuple[str, bytes]:
```